packages/zoe/src/contracts/loan/updateDebt.js

```diff
@@ -30,48 +31,72 @@ export const makeDebtCalculator = debtCalculatorConfig => {
     calcInterestFn = calculateInterest,
     originalDebt,
     loanMath,
-    periodAsyncIterable,
+    periodNotifier,
     interestRate,
+    interestPeriod,
     zcf,
     configMinusGetDebt,
   } = debtCalculatorConfig;
   let debt = originalDebt;
 
+  let lastCalculationUpdate;
+
   const {
     updater: debtNotifierUpdater,
     notifier: debtNotifier,
   } = makeNotifierKit();
 
   const getDebt = () => debt;
 
-  const config = { ...configMinusGetDebt, getDebt };
+  const config = { ...configMinusGetDebt, getDebt, interestPeriod };
 
-  const updateDebt = _state => {
-    const interest = loanMath.make(calcInterestFn(debt.value, interestRate));
-    debt = loanMath.add(debt, interest);
-    debtNotifierUpdater.updateState(debt);
-    scheduleLiquidation(zcf, config);
+  const updateDebt = state => {
+    let prevUpdateTimestamp = lastCalculationUpdate.value;
+    const { value: newTimestamp } = state;
+    while (prevUpdateTimestamp + interestPeriod <= newTimestamp) {
+      prevUpdateTimestamp += interestPeriod;
+      const interest = loanMath.make(calcInterestFn(debt.value, interestRate));
+      debt = loanMath.add(debt, interest);
+      debtNotifierUpdater.updateState(debt);
+    }
   };
 
-  /** @type {IterationObserver<undefined>} */
-  const debtObserver = {
-    // Debt is updated with interest every time the
-    // periodAsyncIterable pushes another value
-    updateState: updateDebt,
-    finish: updateDebt,
-    fail: reason => {
-      debtNotifierUpdater.fail(reason);
-      throw Error(reason);
-    },
-  };
-  harden(debtObserver);
+  function addToDebtWhenNotified(lastCount) {
+    E(periodNotifier)
+      .getUpdateSince(lastCount)
+      .then(
+        newState => {
+          const { updateCount } = newState;
+          if (updateCount) {
+            updateDebt(newState);
+            scheduleLiquidation(zcf, config);
+            lastCalculationUpdate = newState;
+            addToDebtWhenNotified(updateCount);
+          } else {
+            updateDebt(newState);
+            scheduleLiquidation(zcf, config);
+            lastCalculationUpdate = newState;
+          }
+        },
+        reason => {
+          debtNotifierUpdater.fail(reason);
+          throw Error(reason);
+        },
+      );
+  }
 
   // Initialize
-  observeIteration(periodAsyncIterable, debtObserver);
+  E(periodNotifier)
+    .getUpdateSince()
+    .then(update => {
+      lastCalculationUpdate = update;
+      addToDebtWhenNotified(update.updateCount);
+    });
   debtNotifierUpdater.updateState(debt);
 
   return harden({
     getDebt,
+    getLastCalculationTimestamp: _ => lastCalculationUpdate.value,
```

Comment:
Can we use the same notation as the other method?
```suggestion
    getLastCalculationTimestamp: () => lastCalculationTimestamp,
```

Additionally, as this is written, there will be a time before line 92 is called, in which `lastCalculationUpdate` is undefined, and `lastCalculationUpdate.value` will throw a TypeError because it's getting a property on `undefined`. If we instead use `lastCalculationTimestamp` directly, this will return `undefined` without error. 

I wonder if we can improve further. It's a little confusing from the borrower's perspective even with this change, because it will take two changes to `lastCalculationTimestamp` before it indicates any interest in calculated: first, `lastCalculationTimestamp` changes from undefined to a timestamp, but no interest has been calculated. It's only when it changes to the second timestamp that interest has been calculated. It'd be nicer if any change to `lastCalculationTimestamp` indicated that interest had been charged, but I can see that would require waiting for the periodNotifier to give us the first timestamp.

Reply:
I originally tried to address that problem with an early call to `E(periodNotifier).getCurrentTimestamp()`, but that requires an await.

If the client tries to detect the first payment by polling `getLastCalculationTimestamp()`, they'll be sad in the way you identified. If they use the debtNotifier, they should get the answer they want.